File: 3249. Count the Number of Good Nodes.java
class Solution {
    int answer = 0;
    public int countGoodNodes(int[][] edges) {
        Map<Integer, List<Integer>> graph = new HashMap<>();
        for (int[] i : edges) {
            graph.computeIfAbsent(i[0], k -> new ArrayList<>()).add(i[1]);
            graph.computeIfAbsent(i[1], k -> new ArrayList<>()).add(i[0]);
        }
        getCount(graph, 0, -1);
        return answer;
    }
    private int getCount(Map<Integer, List<Integer>> graph, int current, int parent) {
        if (!graph.containsKey(current)) {
            answer++;
            return 1;
        }
        int previous = -1;
        int total = 0;
        boolean flag = true;
        for (int next : graph.get(current)) {
            if (next == parent) continue;
            int count = getCount(graph, next, current);
            if (previous == -1) previous = count;
            if (count != previous) flag = false;
            total += count;
        }
        if (flag) answer++;
        return total + 1;
    }
}

File: go/3249. Count the Number of Good Nodes.go
func countGoodNodes(edges [][]int) int {
	answer := 0
	graph := make(map[int][]int)
	for _, e := range edges {
		if _, ok := graph[e[0]]; !ok {
			graph[e[0]] = []int{}
		}
		if _, ok := graph[e[1]]; !ok {
			graph[e[1]] = []int{}
		}
		graph[e[0]] = append(graph[e[0]], e[1])
		graph[e[1]] = append(graph[e[1]], e[0])
	}
	var getCount func(map[int][]int, int, int) int
	getCount = func(graph map[int][]int, current, parent int) int {
		if _, ok := graph[current]; !ok {
			answer++
			return 1
		}
		previous, total, flag := -1, 0, true
		for _, next := range graph[current] {
			if next == parent { continue }
			count := getCount(graph, next, current)
			if previous == -1 { previous = count }
			if count != previous { flag = false }
			total += count
		}
		if flag { answer++ }
		return total + 1
	}
	getCount(graph, 0, -1)
	return answer
}